cmd/relayer/main.go

```diff
@@ -0,0 +1,412 @@
+	rcfg := &relayer.Config{
+		Ctx:                     ctx,
+		EthClient:               ec,
+		Forwarder:               rcontracts.NewIForwarderWrapped(forwarder),
+		Key:                     key,
+		ValidateTransactionFunc: v.validateTransactionFunc,
+	}
+
+	r, err := relayer.NewRelayer(rcfg)
+	if err != nil {
+		return err
+	}
+
+	if c.Bool(flagWithNetwork) {
+		// cfg.DataDir already has a default set, so only override if the user explicitly set the flag
+		var datadir string
+		if c.IsSet(flagDataDir) {
+			datadir = c.String(flagDataDir) // override the value derived from `flagEnv`
+		} else {
+			datadir = config.DataDir
+		}
+		datadir = path.Join(datadir, "relayer")
+		if err = common.MakeDir(datadir); err != nil {
+			return err
+		}
+
+		h, err := setupNetwork(ctx, c, ec, r, datadir) //nolint:govet
+		if err != nil {
+			return err
+		}
+
+		defer func() {
+			_ = h.Stop()
+		}()
+	}
+
+	if c.Bool(flagRPC) {
+		go signalHandler(ctx, cancel)
+		rpcCfg := &rrpc.Config{
+			Ctx:     ctx,
+			Address: fmt.Sprintf("127.0.0.1:%d", port),
+			Relayer: r,
+		}
+
+		server, err := rrpc.NewServer(rpcCfg) //nolint:govet
+		if err != nil {
+			return err
+		}
+
+		err = server.Start()
+		if errors.Is(err, context.Canceled) || errors.Is(err, http.ErrServerClosed) {
+			return nil
+		}
+	} else {
+		signalHandler(ctx, cancel)
+	}
+
+	return err
+}
+
+func setupNetwork(
+	ctx context.Context,
+	c *cli.Context,
+	ec *ethclient.Client,
+	r *relayer.Relayer,
+	datadir string,
+) (*net.Host, error) {
+	chainID, err := ec.ChainID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var bootnodes []string
+	if c.IsSet(flagBootnodes) {
+		bootnodes = cliutil.ExpandBootnodes(c.StringSlice(flagBootnodes))
+	}
+
+	listenIP := "0.0.0.0"
+	netCfg := &p2pnet.Config{
+		Ctx:        ctx,
+		DataDir:    datadir,
+		Port:       uint16(c.Uint(flagLibp2pPort)),
+		KeyFile:    c.String(flagLibp2pKey),
+		Bootnodes:  bootnodes,
+		ProtocolID: fmt.Sprintf("/%s/%d/%s", swapnet.ProtocolID, chainID.Int64(), net.ProtocolID),
+		ListenIP:   listenIP,
+	}
+
+	cfg := &net.Config{
+		Context:              ctx,
+		P2pConfig:            netCfg,
+		TransactionSubmitter: r,
+		IsRelayer:            true,
+	}
+
+	h, err := net.NewHost(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	err = h.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	return h, nil
+}
+
+func deployOrGetForwarder(
+	addressString string,
+	ec *ethclient.Client,
+	key *rcommon.Key,
+	chainID *big.Int,
+) (*rcontracts.IForwarder, ethcommon.Address, error) {
+	txOpts, err := bind.NewKeyedTransactorWithChainID(key.PrivateKey(), chainID)
+	if err != nil {
+		return nil, ethcommon.Address{}, fmt.Errorf("failed to make transactor: %w", err)
+	}
+
+	if addressString == "" {
+		address, tx, _, err := rcontracts.DeployForwarder(txOpts, ec) //nolint:govet
+		if err != nil {
+			return nil, ethcommon.Address{}, err
+		}
+
+		_, err = bind.WaitMined(context.Background(), ec, tx)
```

Comment:
Don't we have a context that should have been passed in for this?

Reply:
updated